distributed/tests/test_client.py

```diff
@@ -3750,3 +3750,37 @@ def test_dont_clear_waiting_data(c, s, a, b):
     for i in range(5):
         assert s.waiting_data[x.key]
         yield gen.moment
+
+@gen_cluster(client=True)
+def test_recreate_error_simple(c, s, a, b):
+    def makes_error(x):
+        return 1 / x
+
+    f = c.submit(makes_error, 0)
+    yield _wait(f)
+    assert f.status == 'error'
+
+    function, args, kwargs, deps = yield c._get_futures_error(f)
+    assert function.__name__ == 'makes_error'
+    with pytest.raises(ZeroDivisionError):
+        function(*args, **kwargs)
+
+
+@gen_cluster(client=True)
+def test_recreate_error_complex(c, s, a, b):
+    def makes_error(x):
+        return 1 / x
+
+    x = delayed(makes_error)(1)
+    y = delayed(makes_error)(0)
+    tot = delayed(sum)(x, y)
+
+    f = c.compute(tot)
+    yield _wait(f)
+    assert f.status == 'error'
+
+    function, args, kwargs, deps = yield c._get_futures_error(f)
+    assert function.__name__ == 'makes_error'
+    assert args ==  (0,)
+    with pytest.raises(ZeroDivisionError):
+        function(*args, **kwargs)
```

Comment:
Given the use of futures_of above it would be nice to also include an example where a collection fails.  Perhaps something like the following:

```python
b = db.range(10, npartitions=4)
b = b.map(lambda x: 1 / x)
b = c.persist(b)

with pytest.raises(NotImplementedError):
    yield c._recreate_error_locally(b)
```